R/appMetadata.R

```diff
@@ -0,0 +1,378 @@
+appMetadata <- function(appDir,
+                        appFiles = NULL,
+                        appPrimaryDoc = NULL,
+                        quarto = NULL,
+                        contentCategory = NULL,
+                        isCloudServer = FALSE,
+                        metadata = list()) {
+
+  if (is.null(appFiles)) {
+    appFiles <- bundleFiles(appDir)
+  }
+
+  if (!is.null(contentCategory)) {
+    assetTypeName <- contentCategory
+  } else if (!is.null(appPrimaryDoc)) {
+    assetTypeName <- "document"
+  } else {
+    assetTypeName <- "application"
+  }
+
+  quartoInfo <- inferQuartoInfo(
+    appDir = appDir,
+    appPrimaryDoc = appPrimaryDoc,
+    appFiles = appFiles,
+    quarto = quarto,
+    metadata = metadata
+  )
+  appMode <- inferAppMode(
+    appDir = appDir,
+    appPrimaryDoc = appPrimaryDoc,
+    appFiles = appFiles,
+    quartoInfo = quartoInfo,
+    isCloudServer = isCloudServer
+  )
+  appPrimaryDoc <- inferAppPrimaryDoc(
+    appPrimaryDoc = appPrimaryDoc,
+    appFiles = appFiles,
+    appMode = appMode
+  )
+  hasParameters <- appHasParameters(
+    appDir = appDir,
+    appPrimaryDoc = appPrimaryDoc,
+    appMode = appMode,
+    contentCategory = contentCategory
+  )
+  documentsHavePython <- detectPythonInDocuments(
+    appDir = appDir,
+    files = appFiles
+  )
+
+  list(
+    assetTypeName = assetTypeName,
+    appMode = appMode,
+    appPrimaryDoc = appPrimaryDoc,
+    hasParameters = hasParameters,
+    documentsHavePython = documentsHavePython,
+    quartoInfo = quartoInfo
+  )
+}
+
+# infer the mode of the application from its layout
+# unless we're an API, in which case, we're API mode.
+inferAppMode <- function(appDir,
+                         appPrimaryDoc = NULL,
+                         appFiles = NULL,
+                         quartoInfo = NULL,
+                         isCloudServer = FALSE) {
+
+  if (is.null(appFiles)) {
+    appFiles <- bundleFiles(appDir)
+  }
+
+  # TODO(HW): detection rules only apply to files in top-level directory
+  # Should make that explicit in the code
```

Comment:
On the runtime side, we effectively use things like `shiny::runApp()` using the directory defined by the bundle as the working directory. The content can reference resources in sub-directories (as allowed by `shiny` or the in-use framework), but those packages all expect that the primary "thing" is at that top-level.

The one exception to this: If folks should happen to accidentally create wrapping directories when constructing the `bundle.tar.gz`, those get unwound and discarded (e.g. `./wolfgang/amadeus/mozart/app.R` in the archive becomes ./app.R` assuming no other entries at those intermediate levels).

Reply:
Do you think that it would be helpful to document these content-directory expectations in terms of how the code effectively gets run by Connnect (e.g. the `runApp` calls)? I can help enumerate those things, if you'd like.

Reply:
I was meaning something a bit more narrowly scoped. Currently we have code like:

```R
rmdFiles <- grep("^[^/\\\\]+\\.rmd$", appFiles, ignore.case = TRUE, perl = TRUE, value = TRUE)
```

I was thinking we could make the regexps simpler and the logic clearer by doing:

```R
rootFiles <- appFiles[dirname(appFiles) = "."]
# ...
rmdFiles <- grep("\\.rmd$", rootFiles, ignore.case = TRUE, perl = TRUE, value = TRUE)
```

Reply:
Ah. Gotcha. That's a very different TODO! :)